addons/mrp/models/mrp_workorder.py

```diff
@@ -650,29 +650,35 @@ def button_start(self, raise_on_invalid_state=False):
     def button_finish(self):
         date_finished = fields.Datetime.now()
         all_vals_dict = defaultdict(lambda: self.env['mrp.workorder'])
+        to_be_ended_workorders = self.env['mrp.workorder']
+        moves = (self.move_raw_ids + self.production_id.move_byproduct_ids.filtered(lambda m: (m.operation_id == self.operation_id)))
         for workorder in self:
             if workorder.state in ('done', 'cancel'):
                 continue
-            moves = (self.move_raw_ids + self.production_id.move_byproduct_ids.filtered(lambda m: m.operation_id == self.operation_id))
+            moves = moves.filtered(lambda move: not move.picked)
```

Comment:
1) Avoid shadowing variable names, it's common source of bugs. You should always attach some semantic meaning to variable so they describe their content.

2) Moving the `if` in the loop externally is not necessary, since we're looping over the records anyway. You will have a fetch on the first loop, then there are no queries when checking `if not move.picked`. Since we're in stable, patches should have as a minimal diff as possible (to avoid polluting the history of the file when blaming).